### ISSUE TITLE
9.0 web module license is not compatible with the inheritance rules

### ISSUE THREAD

**Issue by gurneyalex** [odoo/odoo#9973]
The fact that the web addon is available under 2 licenses (LGPL in the community version and closed source in the enterprise version) puts the agreement between the OCA and Odoo in an impossible situation, as we cannot reliably port our addons depending explicitly on the web addon (basically all addons in github.com/OCA/web). For instance, we cannot mark the addons as working on the enterprise version easily, and any module depending on web released under AGPL cannot be installed on the enterprise version (when this is the author's will) and we should respect this. 

Can you consider fixing this and for instance rename the modules in the enterprise repository to avoid name clashes?

**Comment by yvaucher** [comment]
:+1: to have separate names

**Comment by rvalyi** [comment]
@yvaucher why not odoo with a o for "open" and clodoo with a "cl" prefix for closed?

**Comment by Danisan** [comment]
Developers can make an LGPL version of their AGPL modules...this reminds me Groucho Marx's quote: "these are my principales, but if you don't like them, I have others"

**Comment by fpodoo** [comment]
**As a technical point of view:**

It's nearly impossible to do the improvements we did in the web client in version 9 on top of the version 8 web framework. In order to make Odoo fully mobile, more responsive, we had to start a lot of things from scratch, including thousands of lines of CSS.

We will, for sure, not do such technical changes for version 9. We might do a common web module with extra inheriting modules for version 10, but we can not guarantee anything for now on. It will mostly depends on our R&D priorities (MRP, Accounting Localisation, bugfixes, ...) and the effort to do it.

Unfortunatelly, we had no technically clean solution to do it in another way for version 9.

**As a license point of view**
I am not a lawyer, nor an expert in these questions, and I did not asked the point of view of our internal experts yet. But here is what I think:

- a developer who develops a module for the community version is in his own rights, even if his module can work on an enterprise version. So, no issue for the developer of a module, whether is license is under AGPL or LGPL.
- the only doubt/discussion could be on the rights of the user: can a user use an AGPL module on a LGPL software having proprietary modules running on top of it? I would say yes because we are not talking about direct dependencies. You can mix any module on the same installation if they don't depend on each others, the same way Linux allows proprietary drivers that were initially developed for Unix. (loadable modules, and not dynamically linked modules)

Of course, the easiest way to deal with that is to release those modules LGPL instead of AGPL. That way, you avoid any conflict or doubt.

**License: a more pragmatic approach**
Having said the above, I also think it's a non issue. It's not about "how do we interpret licenses?", but more about "what do the authors want for their module". My advice to Odoo authors (OCA or others) is to make your intention clear so that there will be no doubts.

- If you want your modules to be compatible with each others, which is better for the end-user and collaboration within the eco-system, just choose a LGPL license.
- If you still want to use AGPL, in my opinion, developers car develop modules that both depend on your module and the web module (from community). If this module also work on the web module of Odoo Enterprise, my opinion is that it's ok for end-users but in this use case, I recommend to make your intentions more explicit in the readme or an official communication to avoid any discussion. [1]
- If you want to be able to sue Odoo users (not developers, you can not in this use case) if they use your modules with Odoo Enterprise, I would suggest to make it more explicit in the license to avoid doubts. Set a license similar to "AGPL+ I don't want you to use my module on an Odoo installation having proprietary modules, even if they don't depend directly on my module".

[1] For what I understood about the OCA position is that they use the AGPL license, with the intention that it will never be a problem for the end-user. And they already made it explicit. If the OCA disagree on this very particular use case, I recommend to make your intentions more explicit that you want to forbid users to use OCA modules in these environments.


Long story short: it's a non issue since you can use LGPL for this very specific use case. This would avoid any doubt. If you want to stick to AGPL, I recommend authors of modules to be explicit about their intention to avoid painful discussion, fears and unnecessary threats on our users;

**Comment by elicoidal** [comment]
@fpodoo Thanks for your reply.
Actually I wanted to make sure there is no misunderstanding.
We do not ask to have the web enterprise released under another license. We respect your choices and this is not the point of this discussion. We understand the RD effort and we **do not ask** for releasing to xGPL the current excellent job in the web enterprise.

As discussed with the OCA, AGPL modules cannot directly depend on closed modules (and vice versa). In order to achieve this we need to have a "web" container module in LGPL from which both OCA modules and the web Enterprise module would depend.
With this "empty" module, we can cleanly inherit the web modules (along with a to-be web_enterprise module): this is a simple technicality, not a game changer request nor a deep structural change (but I am not here the technical guy to judge all consequences).

Without this change, most of OCA web modules become unusable in Enterprise version which is not what Odoo nor OCA is looking for (we have always tried so far to make this "living-together-side-by-side" possible).
I hope this is clearer that way.
--
Eric Caudal [OCA Board member]

**Comment by gurneyalex** [comment]
@fpodoo part of the issue is that there are 2 modules called web, and I'm not sure a module extending web can easily work with both, since the code is so different. So this creates a first problem: we cannot easily say that a module works with web enterprise or with web community. 

Assuming we can create a module which could work for both, we are not free as you seem to think to change its license without getting an "ok" from all the contributors. Such modules could stay AGPL for technical reason beyond our control (and no, the OCA won't force a license on its contributors). As stands, such module would not be installable on top of the enterprise version of Odoo. We need a way to tell this, and currently since there are 2 modules with the same name (web) and 2 different licenses, we cannot do this. 

Thanks for your understanding.

**Comment by odony** [comment]
Hi, 

I'm not sure I understand your problem and what you want to achieve, technically speaking.
The OP apparently mentions 2 problems:

1. "OCA cannot mark the addons as working on the enterprise version easily"
2. "any module depending on web released under AGPL cannot be installed on the enterprise version (when this is the author's will) and we should respect this"

Here are some initial thoughts for each point:

1. The idea with the `web` module of Odoo 9 Enterprise is that it must *replace* and *override* completely the Community web module, in order to avoid very tricky technical problems. And it should also be a mostly-transparent replacement, because both modules share the same API and architecture as much as possible, so other modules should not notice the difference. Theoretically, a module working on the Community `web` module should work on the Enterprise `web` module.
In the future we hope to replace the Community web module with the Enterprise one, and to simplify the situation. But for 9 there's no way around this, as Fabien explained. 
In this context, how do you - technically and practically - suggest we solve your problem? We can't rename the Enterprise `web` module nor relicense it.

2. The problem with "author's will" is that there is no room for such a thing within a pure AGPL world. If you publish an AGPL program, freedom 0 of the [A]GPL grants any user the right to *use* your module however they see fit, you can't forbid any use case.
This is not only the *letter* but also the *intent* of AGPL. The user is not demeaning or damaging your work by using it in combination with a different library. And they're not violating any license either, because they're merely using the different parts, not distributing or modifying them.
This also means there is no need to relicense any of your `web`-dependent modules, they can be used as AGPL on top of either the Community or the Enterprise edition - because they were developed on top of the LGPL `web` API, and that's the only thing they *depend* on.
In essence, this is not different from installing AGPL and Enterprise modules on the same LGPL core - you just happen to have an Enterprise module that more heavily alters the `web` part of the LGPL core, but your AGPL component do not explicitly *depend* on it, in the sense that they would still work without it.

Not sure this answers your concerns, but hopefully we're going a bit further in *some* direction.

**Comment by pedrobaeza** [comment]
But Olivier, that's not what you stated on your license compatibility diagram:

![ir_attachment_5819](https://cloud.githubusercontent.com/assets/7165771/11683171/89f247b6-9e69-11e5-8f9c-2a7eff5dafd6.jpeg)

We can't use an LGPL-3 module that depends on enterprise web module. That's why we ask you to solve this situation for clarifying via dependencies what can be done or not. For example, we have a module that depends on _web_ module. First, we don't know if it's working on both community and enterprise edition only depending on web. Second, assuming the module is working on enterprise, but it's AGPL, technically can be installed, but it shouldn't, or the customer will be on an illegal situation (or at least an uncertain one).

I know that we can put in the README some disclaimer, warning or similar, but we want this to be more easy for everyone to detect incompatibilities or licensing issues.

My proposal is to have a new dummy module called web_enterprise, placed on the enterprise repository, that is auto-installed, and depending only on web.

This way, we can make our OCA modules depending on this new module, and this way, we can highlight that the module is only compatible with the enterprise one. Depending only on web means that you can safely install on both versions of the web module.

This leaves the modules only compatible with community uncovered, but at least we solve some of the possibilities.

What do you think?

**Comment by legalsylvain** [comment]
Hi @odony,

A/ Is there a technical and predictive way to know if the current instance is proprietary or under LGPL licence. I don't see the licence field on manifest file (https://github.com/odoo/odoo/blob/9.0/addons/web/__openerp__.py). (I don't have access to the closed version.)

B/ I see a contradiction between what you said and the graphic below. Don't you ? 

> This also means there is no need to relicense any of your web-dependent modules, they can be used as AGPL on top of either the Community or the Enterprise edition - because they were developed on top of the LGPL web API, and that's the only thing they depend on.


![odoo_agpl](https://cloud.githubusercontent.com/assets/3407482/11683423/f1bab684-9e6a-11e5-9f8f-11b46f62fd7e.jpg)

Thanks a lot for your clarification.

**Comment by odony** [comment]
@pedrobaeza the diagram shows the dependencies, not the deployment. If you make an AGPL module that depends on the LGPL core, deploying it along with the Enterprise `web` module does not change its dependencies. There is no contradiction here. Technically it will run with the enterprise module, but it was not *written* to depend on it. Only developers can violate the license by writing code that mixes incompatible dependencies. Users cannot violate the license by installing modules, as they never *modify* code nor dependencies.

@legalsylvain The license of the Enterprise `web` module is indeed set as `OEEL-1` in the manifest, but my point is that you should not care. Unless your module impacts low-level structures in the web framework and happens to be broken by the Enterprise `web` module. And even then, this might be considered a compatibility bug of the Enterprise edition, if your module uses API entry points that are meant to be extended.

**Comment by gurneyalex** [comment]
@odony I'm not sure geoengine can work on both web clients without significant changes in the code. I could be wrong, though. 

regarding 2, users may perhaps do what they want. But integrators propagate the work, and they are liable.

**Comment by dreispt** [comment]
Maybe this supports @odony 's "not written to depend on" point?:

> __Corresponding Source__ includes interface definition files associated with source files for the work, and the source code for shared libraries and dynamically linked subprograms that __the work is specifically designed to require__, such as by intimate data communication or control flow between those subprograms and other parts of the work.

**Comment by sbidoul** [comment]
I personally share @odony's point of view (https://github.com/odoo/odoo/issues/9973#issuecomment-163177388, item 2).

An AGPL module that depends on the LGPL variant of `web`, and also works on the OE variant, does not *depend* on the OE version: it just happens to works with it too.

If the user decides to use the module on the enterprise version of Odoo I see no issue. In particular, the user's freedom regarding the module is totally preserved since he can decide to stop using the enterprise version while continuing to use the AGPL module.

**Comment by pedrobaeza** [comment]
OK, that clears a lot the scenario when the module is compatible per se with both versions. About geoengine, @gurneyalex, I think the strategy should be to put incompatibilities as issues to be solved by Odoo to assure the same kind of transparency between versions, don't you think?

**Comment by legalsylvain** [comment]
@sbidoul, 
I'm not very confortable with the point of view explained in you last comment. The graphics [like this one](https://cloud.githubusercontent.com/assets/7165771/11683171/89f247b6-9e69-11e5-8f9c-2a7eff5dafd6.jpeg) have been communicated to the whole community during last months.
This graphic indicates that such implementation is not possible. (An AGPL module that depends on a proprietary module).
Accept such a situation doesn't not respect developpers who shared their code under AGPL licence. Don't you think ?

For OCA modules, it is very easy to prevent such situation and leave the choice to the authors to change (or not) the licences of their modules. (As mentionned in the graph, AGPL --> LGPL)

**Comment by sbidoul** [comment]
@legalsylvain I argue my point of view is compatible with the graphic when you interpret the word 'depends' as 'required for the module to function properly', which is a correct interpretation I think. 

I decidedly cannot express it better than @odony (https://github.com/odoo/odoo/issues/9973#issuecomment-163199596).

I personally don't want to restrict the use of ACSONE's AGPL modules on Odoo Enterprise, as long as the user freedom is preserved (hence enforcing the dependency chain on open source only upwards and downwards is important). I also don't want to publish under LGPL because I don't want repackaging rip-offs.

I could understand some authors would want to forbid the use of their modules on Odoo Enterprise, but I don't share that position, as it will lead to more community fragmentation. It would indeed prevent integrators working on Odoo Enterprise to contribute to the `web` repo.

**Comment by rvalyi** [comment]
Hello,

just a word to say I agree with what @pedrobaeza said here https://github.com/odoo/odoo/issues/9973#issuecomment-163233453

I don't think it's ethical at all toward the community, but yes sadly this is probably legal. This is exactly because I was anticipating all the nasty dark trading with the commercial conditions for using the Enterprise blobs and how this is used behind the curtains to try having the proprietary side taking over the open source players that I had a pretty radical stance toward the CLA. But now you have it and these stuff are not illegal anymore in my view; welcome to the capitalism.

**Comment by yvaucher** [comment]
@odony position makes sense, while I wonder as @gurneyalex what is the responsibility of the integrator there.

However, I think it can be harmful for the enterprise version of web module. This because AGPL modules will be developed focusing the community version which mean it might not work properly on enterprise version.

What is true nevertheless is that AGPL is there to protect the freedom of the user. Thus he must be free to use it the way he needs. Still, it will come with less warranty for enterprise version.

**Comment by legalsylvain** [comment]
Hi all,

Thanks a lot for all your explainations. I really didn't understood 

If I try to sum up :

* **Case 1** : Implementation on Odoo Enterprise (V9).

A module 'oca_extra_accounting_feature' / (by OCA / AGPL) is installed (and depends on) 'extra_private_accounting_feature' / (by Odoo SA / OEEL-1).

-> This is forbidden. [Ref](http://image.slidesharecdn.com/2015-jodoo-150616152909-lva1-app6891/95/jornadas-odoo-2015-charla-con-odoo-s-a-31-638.jpg?cb=1434480500).

* **Case 2** :  Implementation on Odoo Community Version (V9).A new module 'oca_web_extra_feature' / (by OCA / AGPL) is installed (and depends on) 'web' (Odoo SA / LGPL).

-> This is allowed. [Ref](http://image.slidesharecdn.com/2015-jodoo-150616152909-lva1-app6891/95/jornadas-odoo-2015-charla-con-odoo-s-a-32-638.jpg?cb=1434480500).



Case 3/ Implementation on Odoo Enterprise (V9).
A new module 'oca_web_extra_feature' / (by OCA / AGPL) is installed (and depends on) 'web' (Odoo SA / OEEL-1).

-> This is allowed because implementation 2 exists and works even if the current 'web' module is under a proprietary license.



Did I understood correctly, or did I miss something ?

If I'm wrong, please explain in which case [this configuration](http://image.slidesharecdn.com/2015-jodoo-150616152909-lva1-app6891/95/jornadas-odoo-2015-charla-con-odoo-s-a-31-638.jpg?cb=1434480500) exists.



full ref doc : http://fr.slideshare.net/pedromanuelbaezaromero/jornadas-odoo-2015-49463544

**Comment by Danisan** [comment]
So, if I have "module_a" (AGPL) developed for 8.0. This is my fork of "module_z" (agpl), very tuned, very elaborated, an very very changed, but with pieces of code from module_z yet... Suppose module_a is more popular than module_z, (but remember, pieces of module_z's code are still there)....may I create a "module_a LGPL for 9.0? Where is the breaking point to entitle me to do this? in a percentage of the code? what do you think?

**Comment by gurneyalex** [comment]
@Danisan you are not allowed to do this unless you have the agreement from the people who wrote the code you copied from module_z.

**Comment by Yenthe666** [comment]
I believe everything was answered and discussed here. 😄
